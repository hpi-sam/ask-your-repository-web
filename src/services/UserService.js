// @flow
import humps from 'humps';
import api from '../config/api';
import type { User, UserParams } from '../models/User';

class UserService {
  static async get(id: string): Promise<User> {
    const response = await api.get(`/users/${id}`);
    return response.data;
  }

  static async list(): Promise<User[]> {
    const response = await api.get('/users', {});

    return response.data;
  }

  static async create(formData: UserParams): Promise<User> {
    try {
      const response = await api.post('/users', formData);
      return response.data;
    } catch (e) {
      throw e;
    }
  }

  static async login(emailOrUsername: string, password: string): Promise<User> {
    try {
      const response = await api.post('/authentications', humps.decamelizeKeys({ emailOrUsername, password }));
      return response.data;
    } catch (e) {
      throw e;
    }
  }

  static async logout() {
    try {
      await api.delete('/authentications');
    } finally {
      localStorage.removeItem('user');
    }
  }
}

export default UserService;
