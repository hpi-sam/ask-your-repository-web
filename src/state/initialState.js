// @flow

// This file is only for state related testing

import type { AppState } from './AppState';
import { initialState as presentation } from './presentation/presentation.reducer';
import { initialState as teamSidebar } from './team_sidebar/teamSidebar.reducer';
import { initialState as presentationMode } from './presentation_mode/presentationMode.reducer';
import { initialState as activeTeam } from './active_team/activeTeam.reducer';
import { initialState as auth } from './auth/auth.reducer';

const initialState: AppState = {
  flash: { messages: [] },
  presentation,
  activeTeam,
  teamSidebar,
  presentationMode,
  auth,
};

export default initialState;
