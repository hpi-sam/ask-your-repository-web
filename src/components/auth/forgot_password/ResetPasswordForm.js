// @flow
import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { push } from 'connected-react-router';
import humps from 'humps';
import qs from 'qs';
import { Link } from 'react-router-dom';
import { flashErrorMessage, flashSuccessMessage } from 'redux-flash';
import Form from 'react-validation/build/form';
import Input from 'react-validation/build/input';
import Button from 'react-validation/build/button';
import UserService from '../../../services/UserService';
import ValidationErrors from '../../utility/form/ValidationErrors';
import type { Errors } from '../../../models/Errors';

type Props = {
  location: Object,
};

const ResetPasswordForm = (props: Props) => {
  const dispatch = useDispatch();
  const [password, setPassword] = useState('');
  const [passwordConfirm, setPasswordConfirm] = useState('');
  const [errors, setErrors] = useState<Errors>({
    missingInput: false,
    mismatchedPassword: false,
    invalidEmail: false,
  });

  const resetErrors = () => {
    setErrors({
      missingInput: false,
      mismatchedPassword: false,
      invalidEmail: false,
    });
  };

  const handleSubmit = async (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();
    resetErrors();

    if (password && passwordConfirm) {
      if (password !== passwordConfirm) {
        setErrors({ ...errors, mismatchedPassword: true });
        return false;
      }

      const querystring = props.location.search;
      const parsedString = qs.parse(querystring, { ignoreQueryPrefix: true });
      const { resetToken } = humps.camelizeKeys(parsedString);
      const token: any = resetToken;

      try {
        await UserService.changePassword(password, token);
        dispatch(flashSuccessMessage('Successfully updated the password.'));
        dispatch(push('/login'));
      } catch (error) {
        const message = error.response
          ? error.response.data.error
          : 'Could not establish a connection to the server.';

        dispatch(flashErrorMessage(message));
      }
    } else {
      setErrors({ ...errors, missingInput: true });
    }
    return true;
  };

  const printError = () => {
    const error = Object.keys(errors).find(key => errors[key]);
    return error
      ? <ValidationErrors error={error} />
      : false;
  };

  return (
    <Form onSubmit={handleSubmit} className="Form Form__centered">
      <div className="Form__title">Update Password</div>
      {printError()}
      <div className="Form__input">
        <label className="Form__input__label">
          Password:
          <Input
            type="password"
            name="password"
            value={password}
            onChange={e => setPassword(e.target.value)}
            data-cy="register-password-input"
          />
        </label>
      </div>
      <div className="Form__input">
        <label className="Form__input__label">
          Confirm Password:
          <Input
            type="password"
            name="passwordConfirm"
            value={passwordConfirm}
            onChange={e => setPasswordConfirm(e.target.value)}
            data-cy="register-password-repeat-input"
          />
        </label>
      </div>

      <div className="Form__buttons">
        <Button
          className="Form__buttons__item Form__buttons__item__blue"
          data-cy="register-submit-button"
        >
          Update Password
        </Button>
        <Link
          to="/"
          className="Form__buttons__item Form__buttons__item__gray"
        >
          Cancel
        </Link>
      </div>
    </Form>
  );
};

export default ResetPasswordForm;
