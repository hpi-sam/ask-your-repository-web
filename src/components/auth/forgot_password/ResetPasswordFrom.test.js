// @flow
import React from 'react';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import Form from 'react-validation/build/form';
import ResetPasswordForm from './ResetPasswordForm';
import initialState from '../../../state/initialState';
import ValidationErrors from '../../utility/form/ValidationErrors';
import UserService from '../../../services/UserService';

const mockStore = configureMockStore();

jest.mock('../../../services/UserService');

describe('<ResetPasswordForm />', () => {
  let wrapper;
  let store;

  beforeAll(() => {
    UserService.changePassword.mockImplementation(() => Promise.resolve([]));
  });

  beforeEach(() => {
    store = mockStore(initialState);
    const location = { search: '?reset_token=test_token' };
    wrapper = mount((
      <Provider store={store}>
        <MemoryRouter>
          <ResetPasswordForm location={location} />
        </MemoryRouter>
      </Provider>
    )).find('ResetPasswordForm');
  });

  afterEach(() => {
    UserService.changePassword.mockClear();
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('shows a validation Error when submitting without providing a password', () => {
    wrapper.find(Form).simulate('submit');
    wrapper = wrapper.update();
    expect(wrapper.find(ValidationErrors).text()).toEqual('Please fill in all fields.');
  });

  it('shows a validation Error when submitting with mismatching passwords', () => {
    wrapper.find('input[name="password"]')
      .simulate('change', { target: { name: 'password', value: 'test' } });
    wrapper.find('input[name="passwordConfirm"]')
      .simulate('change', { target: { name: 'password', value: 'mismatched_password' } });
    wrapper.find(Form).simulate('submit');
    wrapper = wrapper.update();
    expect(wrapper.find(ValidationErrors).text()).toEqual('Passwords do not match.');
  });

  it('calls the auth service with the password', () => {
    wrapper.find('input[name="password"]')
      .simulate('change', { target: { name: 'password', value: 'test' } });
    wrapper.find('input[name="passwordConfirm"]')
      .simulate('change', { target: { name: 'password', value: 'test' } });
    wrapper.find(Form).simulate('submit');
    expect(UserService.changePassword).toHaveBeenCalledWith('test', 'test_token');
  });
});
