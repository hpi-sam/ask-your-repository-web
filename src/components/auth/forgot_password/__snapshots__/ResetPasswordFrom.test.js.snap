// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetPasswordForm /> renders correctly 1`] = `
<ResetPasswordForm
  location={
    Object {
      "search": "?reset_token=test_token",
    }
  }
>
  <Form(t)
    className="Form Form__centered"
    onSubmit={[Function]}
  >
    <t
      className="Form Form__centered"
      getValues={[Function]}
      hideError={[Function]}
      onSubmit={[Function]}
      showError={[Function]}
      validate={[Function]}
      validateAll={[Function]}
    >
      <form
        className="Form Form__centered"
        onSubmit={[Function]}
      >
        <div
          className="Form__title"
        >
          Update Password
        </div>
        <div
          className="Form__input"
        >
          <label
            className="Form__input__label"
          >
            Password:
            <Control(s)
              data-cy="register-password-input"
              name="password"
              onChange={[Function]}
              type="password"
              validations={Array []}
              value=""
            >
              <s
                data-cy="register-password-input"
                name="password"
                onBlur={[Function]}
                onChange={[Function]}
                type="password"
                value=""
              >
                <div>
                  <input
                    data-cy="register-password-input"
                    name="password"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="password"
                    value=""
                  />
                </div>
              </s>
            </Control(s)>
          </label>
        </div>
        <div
          className="Form__input"
        >
          <label
            className="Form__input__label"
          >
            Confirm Password:
            <Control(s)
              data-cy="register-password-repeat-input"
              name="passwordConfirm"
              onChange={[Function]}
              type="password"
              validations={Array []}
              value=""
            >
              <s
                data-cy="register-password-repeat-input"
                name="passwordConfirm"
                onBlur={[Function]}
                onChange={[Function]}
                type="password"
                value=""
              >
                <div>
                  <input
                    data-cy="register-password-repeat-input"
                    name="passwordConfirm"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="password"
                    value=""
                  />
                </div>
              </s>
            </Control(s)>
          </label>
        </div>
        <div
          className="Form__buttons"
        >
          <Button(s)
            className="Form__buttons__item Form__buttons__item__blue"
            data-cy="register-submit-button"
          >
            <s
              className="Form__buttons__item Form__buttons__item__blue"
              data-cy="register-submit-button"
              hasErrors={false}
            >
              <button
                className="Form__buttons__item Form__buttons__item__blue"
                data-cy="register-submit-button"
                disabled={false}
              >
                Update Password
              </button>
            </s>
          </Button(s)>
          <Link
            className="Form__buttons__item Form__buttons__item__gray"
            to="/"
          >
            <a
              className="Form__buttons__item Form__buttons__item__gray"
              href="/"
              onClick={[Function]}
            >
              Cancel
            </a>
          </Link>
        </div>
      </form>
    </t>
  </Form(t)>
</ResetPasswordForm>
`;
